packages/runtime/container-runtime/src/dataStores.ts

```diff
@@ -0,0 +1,243 @@
+            attachMessage.id,
+            snapshotTreeP,
+            this.runtime,
+            new BlobCacheStorageService(this.runtime.storage, flatBlobsP),
+            this.runtime.scope,
+            this.summaryTracker.createOrGetChild(attachMessage.id, message.sequenceNumber),
+            this.getCreateChildSummarizerNodeFn(
+                attachMessage.id,
+                {
+                    type: CreateSummarizerNodeSource.FromAttach,
+                    sequenceNumber: message.sequenceNumber,
+                    snapshot: attachMessage.snapshot ?? {
+                        id: null,
+                        entries: [createAttributesBlob(pkg, true /* isRootDataStore */)],
+                    },
+                }),
+            pkg);
+
+        // Resolve pending gets and store off any new ones
+       this.contexts.addBoundOrRemote(attachMessage.id, remotedFluidDataStoreContext);
+
+        // Equivalent of nextTick() - Prefetch once all current ops have completed
+        // eslint-disable-next-line @typescript-eslint/no-floating-promises
+        Promise.resolve().then(async () => remotedFluidDataStoreContext.realize());
+    }
+
+    public  bindFluidDataStore(fluidDataStoreRuntime: IFluidDataStoreChannel): void {
+        const id = fluidDataStoreRuntime.id;
+        const localContext = this.contexts.prepContextForBind(id);
+        // If the container is detached, we don't need to send OP or add to pending attach because
+        // we will summarize it while uploading the create new summary and make it known to other
+        // clients.
+        if (this.runtime.attachState !== AttachState.Detached) {
+            localContext.emit("attaching");
+            const message = localContext.generateAttachMessage();
+
+            this.pendingAttach.set(id, message);
+            this.submitAttachFn(message);
+            this.attachOpFiredForDataStore.add(id);
+        }
+
+        // Resolve the deferred so other local stores can access it now that the context is bound
+        this.contexts.resolveDeferredBind(fluidDataStoreRuntime.id);
+    }
+
+    public createDetachedDataStoreCore(
+        pkg: Readonly<string[]>,
+        isRoot: boolean,
+        id = uuid()): IFluidDataStoreContextDetached
+    {
+        const context = new LocalDetachedFluidDataStoreContext(
+            id,
+            pkg,
+            this.runtime,
+            this.runtime.storage,
+            this.runtime.scope,
+            this.summaryTracker.createOrGetChild(id, this.runtime.deltaManager.lastSequenceNumber),
+            this.getCreateChildSummarizerNodeFn(id, { type: CreateSummarizerNodeSource.Local }),
+            (cr: IFluidDataStoreChannel) => this.bindFluidDataStore(cr),
+            undefined,
+            isRoot,
+        );
+        this.contexts.addUnbound(context);
+        return context;
+    }
+
+    public _createFluidDataStoreContext(pkg: string[], id: string, isRoot: boolean, props?: any) {
```

Comment:
So these 2 can be public because dataStores.ts isn't exported via index.ts?

Reply:
yup.